libnetwork/default_gateway.go

```diff
@@ -175,15 +175,31 @@ func (c *Controller) defaultGwNetwork() (*Network, error) {
 	return n, err
 }
 
-// Returns the endpoint which is providing external connectivity to the sandbox
-func (sb *Sandbox) getGatewayEndpoint() *Endpoint {
-	for _, ep := range sb.Endpoints() {
+// getGatewayEndpoint returns the endpoints providing external connectivity to
+// the sandbox. If the gateway is dual-stack, ep4 and ep6 will point at the same
+// endpoint. If there is no IPv4/IPv6 connectvity, nil pointers will be returned.
+func (sb *Sandbox) getGatewayEndpoint() (ep4, ep6 *Endpoint) {
+	return selectGatewayEndpoint(sb.Endpoints())
+}
+
+// selectGatewayEndpoint is like getGatewayEndpoint, but selects only from
+// endpoints. It can be called when the Sandbox mutex is already held.
```

Comment:
I split `selectGatewayEndpoint` out of `getGatewayEndpoint` so that it could be called from `clearNetworkResources` with the mutex already held.

`getGatewayEndpoint` now calls `selectGatewayEndpoint(sb.Endpoints())` ... and `sb.Endpoints()` acquires the mutex to make a copy of the slice.

Previously, `getGatewayEndpoint` did this, which is equivalent ...

```
func (sb *Sandbox) getGatewayEndpoint() *Endpoint {
	for _, ep := range sb.Endpoints() {
		...
```

So, if there's a path with a race, I don't think this change affects it.